### ISSUE TITLE
ThreadVectorRange parallel_reduce (int, +) gives incorrect result on SYCL

### ISSUE THREAD

**Issue by brian-kelley** [kokkos/kokkos#4573]
I have a parallel_reduce with ThreadVectorRange that is used to count the unique integers in a sorted array, but on SYCL this reduction returns a higher value than what is correct. Here's a reproducer:
```
#include <Kokkos_Core.hpp>
#include <algorithm>
#include <cstdlib>

using ExecSpace = Kokkos::DefaultExecutionSpace;
using Device = Kokkos::Device<ExecSpace, typename ExecSpace::memory_space>;

using IntView = Kokkos::View<int*, Device>;

struct TestFunctor {
  using TeamPol = Kokkos::TeamPolicy<ExecSpace>;
  using TeamMem = typename TeamPol::member_type;

  TestFunctor(const IntView& v_) : v(v_) {}

  KOKKOS_INLINE_FUNCTION void operator()(const TeamMem t) const {
    int risingEdges = 0;
    int n = v.extent_int(0);
    Kokkos::parallel_reduce(
        Kokkos::ThreadVectorRange(t, n - 1), 
        [=](const int j, int& lcount) {
          if (v(j) != v(j + 1)) lcount++;
        },  
        risingEdges);
    Kokkos::single(Kokkos::PerThread(t),
    [=]()
    {   
      printf("parallel_reduce counted %d unique values in view.\n", risingEdges + 1); 
      int trueUnique = 1;
      for(int i = 1; i < n; i++)
      {   
        if(v(i - 1) != v(i))
          trueUnique++;
      }   
      printf("From device: true #unique is %d\n", trueUnique);
    }); 
  }

  IntView v;
};

int main() {
  Kokkos::initialize();
  {
    int n = 1234;
    IntView v("v", n); 
    auto vhost = Kokkos::create_mirror_view(v);
    //Generate a monotonically increasing list with some duplicates
    for(int i = 1; i < n; i++)
    {   
      int incr = rand() % 2;
      vhost(i) = vhost(i - 1) + incr;
    }   
    Kokkos::deep_copy(v, vhost);
    std::cout << "Max vector length: " << TestFunctor::TeamPol::vector_length_max() << '\n';
    Kokkos::parallel_for(typename TestFunctor::TeamPol(1, 1, 32), TestFunctor(v));
    //Now see true number of unique values in view
    int trueUnique = 1;
    for(int i = 1; i < n; i++)
    {   
      if(vhost(i - 1) != vhost(i))
        trueUnique++;
    }
    std::cout << "Functor above should have printed " << trueUnique << '\n';
    int* vhostBegin = vhost.data();
    int* vhostEnd = vhost.data() + vhost.extent(0);
    std::cout << "One extra check using std::unique: " << std::unique(vhostBegin, vhostEnd) - vhostBegin << '\n';
  }
  Kokkos::finalize();
  return 0;
}
```
The vector_length_max is reported to be 32, but using that in the policy causes the bug. If I plug in 8 or 16 as the vector_length instead, it gives the correct result. A device-level RangePolicy reduce also worked, but I didn't include that here.

Example output showing the bug (the order of the lines can get mixed up, since the device printouts don't get flushed right away):
```
Max vector length: 32
parallel_reduce counted 639 unique values in view.
Functor above should have printed 618
One more extra check using std::unique: 618
From device: true #unique is 618
```

**Comment by masterleinad** [comment]
In `Debug` mode, there should be an error message that the maximum vector length for this kernel is 8 (but for some reason, it is not printed). Unfortunately, it is quite tricky to figure out the actual maximum vector size for a kernel at compile-time (instead of the maximum one supported by the device).

**Comment by brian-kelley** [comment]
@masterleinad It looks like vector_length_max is a runtime computation, but it obviously doesn't know which kernel you're going to run. Can different kernels have different maximums within the same build?
```
  static int vector_length_max() {
    std::vector<size_t> sub_group_sizes =
        execution_space{}
            .impl_internal_space_instance()
            ->m_queue->get_device()
            .template get_info<sycl::info::device::sub_group_sizes>();
    return *std::max_element(sub_group_sizes.begin(), sub_group_sizes.end());
  }
```

**Comment by masterleinad** [comment]
https://github.com/kokkos/kokkos/pull/4579 tries to improve the detection of compatibility between subgroup size and vector size but it's blocked by https://github.com/intel/llvm/issues/5071.

**Comment by masterleinad** [comment]
For the moment, the check prints something if you compile with `-DNDEBUG` (`Release` mode) and `Kokkos_ENABLE_DEBUG=ON`. I'm not quite sure what else to do about it at the moment. :shrug:

**Comment by brian-kelley** [comment]
@masterleinad That's totally fine, thanks. I'm not in a hurry to get 100% of stuff working with SYCL, I'm just looking for issues to make sure somebody's aware of them.

**Comment by masterleinad** [comment]
@brian-kelley Would you mind checking if #4704 (assuming that it build and passes `Kokkos` unit tests) allows running the `KokkosKernels` unit tests without using the workarounds for the vector size?

**Comment by brian-kelley** [comment]
@masterleinad This does work everywhere in KokkosKernels, except spgemm because it actually does explicitly use the vector length, even though it is a parallel_for:
```
    int vector_rank = 0; 
    typedef typename std::remove_reference<decltype(*used_hash_sizes)>::type
        atomic_incr_type;
    Kokkos::parallel_scan(
        Kokkos::ThreadVectorRange(teamMember, vector_size),
        [&](const int /* threadid */, int &update, const bool final) {
          if (final) {
            vector_rank = update;
          }    
          update += 1;
        });  
    int bs           = vector_size * suggested_team_size;
    int vector_shift = thread_rank * vector_size + vector_rank;
    ...
```
So this runs into errors because the vector length computed ahead of time (with ``TeamPolicy::vector_length_max()``) and the one your PR computes at launch time don't match. There isn't a way to get the vector length from a TeamMember. Is there some other way to extract the ``final_vector_size`` you compute, without actually launching? If not I can just keep the hardcoded maximum of 8 just for these kernels. This spgemm code is going to be replaced this year anyway.

**Comment by masterleinad** [comment]
We could probably dicuss adding a `vector_rank()` and `vector_size()` member function to the `*TeamMember` classes.

**Comment by crtrott** [comment]
The above thing is not valid Kokkos code as I have stated before. It is a hack which assumes a specific way of how a backend works. The above has no chance of working with OpenMP, Threads, HPX or OpenMPTarget, because it relies on vector lanes behaving like threads (i.e. all vector lanes being active outside the ThreadVectorRange, and the number of vector lanes being determistic).

**Comment by masterleinad** [comment]
Fixed in #4704.

**Comment by PhilMiller** [comment]
Fix released in 3.6
